### DISCUSSION TITLE
Ganache fork polygon doesn't give the right balanceOf result

### DISCUSSION THREAD

**Discussion by lingmicha** [ConsenSys-archive/ganache#3837] [Q&A] [answered]
Hello, I tried to fork a polygon mainnet at block number: 34581439. 

the cmd is as follows:
`ganache-cli --chain.vmErrorsOnRPCResponse true --server.port 19999 --fork.url https://polygon-mainnet.g.alchemy.com/v2/H3E99oH7XymptRueBxkO4OXXazhFb7_c@34581439`

after the local ganache fork is running, i tried to access a aave atoken balance: 

`curl -H "Content-Type: application/json" -X POST --data \
        '{"id":1337,"jsonrpc":"2.0","method":"eth_call","params":[{"from":null, "to":"0x82e64f49ed5ec1bc6e43dad4fc8af9bb3a2312ee","data":"0x70a082310000000000000000000000005b8effbdae5941b938c1a79616d185cc3c79d4ff" }]}' \
        http://localhost:19999
`
But it give a increasing balance that is not fixed at block 34581439. Everytime i call the eth_call, the return figure increases. And it is also not the lastest_block value. That buffles me a lot! 

Anyone can point me to the right direction why this happens?

**Comment by davidmurdoch** [comment]
What version of ganache are you using?

**Reply by lingmicha** [reply]
Ganache: 
ganache v7.4.3 (@ganache/cli: 0.5.3, @ganache/core: 0.5.3)

macOS version: Big Sur: 11.6.1
node version: tried 12.0.0 / 16.0.0

**Reply by davidmurdoch** [reply]
Can you try the latest version, 7.5.0, and let me know if it works for you. I was unable to reproduce the issue earlier today.

**Reply by lingmicha** [reply]
I tried with Ganache 7.5.0. The issue remains. 
I might mis-present the issue previously, let me reiterate: 

Step 1: Start ganache forking polygon mainnet@34581439: 
`ganache-cli --chain.vmErrorsOnRPCResponse true --server.port 19999 --fork.url https://polygon-mainnet.g.alchemy.com/v2/alchemy_key@34581439`
then i call the balanceOf function, get the return value: `0x00000000000000000000000000000000000000000000309b64a51788682cd24f`

Step 2: restart ganache with the same cmd, request again, get the return value. Everytime i did this, the value keep increasing:
`0x00000000000000000000000000000000000000000000309b64a51788682cd24f`
`0x00000000000000000000000000000000000000000000309b674e2a546727c161`
`0x00000000000000000000000000000000000000000000309b67868274e26d3030`


Step 3: I request the actual value on mainnet on that block height, the value should be( and is stable at): 
`0x00000000000000000000000000000000000000000000309b7beddbeba2cec20e`

The contract address: 
https://polygonscan.com/address/0x82e64f49ed5ec1bc6e43dad4fc8af9bb3a2312ee
The read function: balanceOf
The wallet address: 0x5b8effbdae5941b938c1a79616d185cc3c79d4ff


Can you help investigate why this is happening? I need the actual value stable to craft some unit test cases.

Thanks,
Michael

**Reply by lingmicha** [reply]
Doesn't seem to have the same issue when i fork bnb chain. Possibly the archive node issue?

**Reply by davidmurdoch** [reply]
The contract is time-dependent. Ganache currently defaults to using your system time by default, both for mining and simulating blocks.

You can change this by setting `--chain.time = <timestamp>` (where `<timestamp>` is in millisecond precision - like JavaScript's `Date.now()`) -- this sets Ganache's starting time. I'd also recommend setting `--miner.timestampIncrement = 1`, which tells ganache to ignore your system time altogether, as you'll likely still run into time discrepancies if you don't.

In the future ganache will use the fork's blocktime to set out internal clock time offset (which is what `--chain.time` is doing), as well as default to `--miner.timestampIncrement = 1` (it currently defaults to "clock"). It is a breaking change to do so, which is why it isn't the default already.

Sorry for the confusing these default settings have caused! Let me know if that sorts things out for you.

Here is the issue for setting the clock time offset to the fork block's time: https://github.com/trufflesuite/ganache/issues/3186#issuecomment-1177963245

And the issue for the timestampIncrement option: https://github.com/trufflesuite/ganache/issues/3845

**Reply by lingmicha** [reply]
Thank you david！ That explains a lot. 

Let me try it and get back to you

**Reply by lingmicha** [reply]
Hi David, 

Here's what I did and the issue i encountered: 
1) I looked up the block timestamp: Oct-20-2022 02:51:49 PM +UTC
2) Then I start ganache with :
`ganache-cli --chain.vmErrorsOnRPCResponse true --server.port 19998 --fork.url https://polygon-mainnet.g.alchemy.com/v2/H3E99oH7XymptRueBxkO4OXXazhFb7_c@34581439 --hardfork london --chain.time 2022-10-20T14:51:49 --miner.timestampIncrement 1`
3) when i call the balanceOf function, i got an error VM Exception. 

Request: 
`curl -H "Content-Type: application/json" -X POST --data \
'{"id":1337,"jsonrpc":"2.0","method":"eth_call","params":[{"from":null, "to":"0x82e64f49ed5ec1bc6e43dad4fc8af9bb3a2312ee","data":"0x70a082310000000000000000000000005b8effbdae5941b938c1a79616d185cc3c79d4ff" }]}' \
http://localhost:19998`

Return:
`{"id":1337,"jsonrpc":"2.0","error":{"message":"VM Exception while processing transaction: revert","stack":"CallError: VM Exception while processing transaction: revert\n    at Blockchain.simulateTransaction (/home/ubuntu/.nvm/versions/node/v12.22.12/lib/node_modules/ganache/dist/node/1.js:2:81107)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)","code":-32000,"name":"CallError","data":"0x4e487b710000000000000000000000000000000000000000000000000000000000000011"}}`

Did i do something wrong?

**Reply by davidmurdoch** [reply]
`<timestamp>` is in millisecond precision - like JavaScript's `Date.now()`. So:

```
ganache-cli --chain.vmErrorsOnRPCResponse true --server.port 19998 --fork.url https://polygon-mainnet.g.alchemy.com/v2/H3E99oH7XymptRueBxkO4OXXazhFb7_c@34581439 --hardfork london --chain.time 1666291909000 --miner.timestampIncrement 1
```

**Reply by davidmurdoch** [reply]
Actually, a Date string like that should work :thinking:

**Reply by lingmicha** [reply]
I tried the millisecond presentation, the problem stays the same.

Seems to me some inner conflicts introduced by these to options. Without those two, it works.

Can you reproduce the problem using my cmds?

**Reply by davidmurdoch** [reply]
I cannot reproduce. I'm running:

```console
$ ganache-cli --version
ganache v7.5.0 (@ganache/cli: 0.6.0, @ganache/core: 0.6.0)

$ ganache-cli --chain.vmErrorsOnRPCResponse true --server.port 19998 --fork.url https://polygon-mainnet.g.alchemy.com/v2/H3E99oH7XymptRueBxkO4OXXazhFb7_c@34581439 --hardfork london --chain.time 2022-10-20T14:51:49 --miner.timestampIncrement 1
```

and 

```console
$ curl -H "Content-Type: application/json" -X POST --data '{"id":1337,"jsonrpc":"2.0","method":"eth_call","params":[{"from":null, "to":"0x82e64f49ed5ec1bc6e43dad4fc8af9bb3a2312ee","data":"0x70a082310000000000000000000000005b8effbdae5941b938c1a79616d185cc3c79d4ff" }]}' http://localhost:19998
{"id":1337,"jsonrpc":"2.0","result":"0x0000000000000000000000000000000000000000000030991a05e26859c1a032"}
```

I always get the same result, `0x0000000000000000000000000000000000000000000030991a05e26859c1a032`, no matter how many times I restart ganache and run the curl command.

Note: you can now use the command `ganache`, instead of `ganache-cli`.

**Reply by lingmicha** [reply]
Something is weird. 

I tried the same command on my mac and my ubuntu server. Every time, i got a VM Error:

`{"id":1337,"jsonrpc":"2.0","error":{"message":"VM Exception while processing transaction: revert","stack":"CallError: VM Exception while processing transaction: revert\n    at Blockchain.simulateTransaction (/home/ubuntu/.nvm/versions/node/v12.22.12/lib/node_modules/ganache/dist/node/1.js:2:81107)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)","code":-32000,"name":"CallError","data":"0x4e487b710000000000000000000000000000000000000000000000000000000000000011"}}`

But when i start ganache w/o `--chain.time 2022-10-20T14:51:49 --miner.timestampIncrement 1`, the return is valid. 

How could this happen? I tried to debug in many ways, but there's little useful info i can extract from ganache. 

The ganache version: 
On Untutu: `ganache v7.4.4 (@ganache/cli: 0.5.4, @ganache/core: 0.5.4)`
On Mac: `ganache v7.5.0 (@ganache/cli: 0.6.0, @ganache/core: 0.6.0)`

Pls help investigate. Thanks,
Michael

**Reply by lingmicha** [reply]
@davidmurdoch Hi david, i've tried the cmd on various machines. ALL failed with the same response. 

Tried with different ganache version + different node versions. 

I wonder why yours works. :)

**Reply by davidmurdoch** [reply]
Can you try quoting the the time? Maybe whatever shell you are using is attempting to parse it?

**Comment by tenthirtyone** [comment, marked as answer]
@lingmicha I think @davidmurdoch and I figured this one out. 

In our testing, we were both consistently receiving different results. Then we realized, geographically we are in different locations. When we double checked our timestamps, they were the same. However, because we reside in different geographic locations, those timestamps meant different things to our system time. When we passed in a timezone with the date param, our results were consistently the same.

> Timezones

Try to pass the following command to start ganache:

```
$ ganache-cli --chain.vmErrorsOnRPCResponse true --server.port 19998 --fork.url https://polygon-mainnet.g.alchemy.com/v2/H3E99oH7XymptRueBxkO4OXXazhFb7_c@34581439 --hardfork london --chain.time 2022-10-20T14:51:49Z --miner.timestampIncrement 1 --fork.disableCache -d
```

There is one small difference in this command from your previous commands:

`--chain.time 2022-10-20T14:51:49Z`

The `Z` in timezone tells Node to default to UTC time. 

Give this a shot and let us know how it goes.

Edit: Oh, I also ran into the error you shared. When I was swapping API keys david pointed out I forgot to include the block number. Starting without the block number after previous testing caused this error.

STR:
``` sh
ganache-cli --chain.vmErrorsOnRPCResponse true --server.port 19998 --fork.url https://polygon-mainnet.g.alchemy.com/v2/[YOUR KEY HERE]@34581439 --hardfork london --chain.time 2022-10-20T14:51:49Z --miner.timestampIncrement 1
```

``` sh
curl -H "Content-Type: application/json" -X POST --data '{"id":1337,"jsonrpc":"2.0","method":"eth_call","params":[{"from":null, "to":"0x82e64f49ed5ec1bc6e43dad4fc8af9bb3a2312ee","data":"0x70a082310000000000000000000000005b8effbdae5941b938c1a79616d185cc3c79d4ff" }]}' http://localhost:19998                                                     

{"id":1337,"jsonrpc":"2.0","result":"0x000000000000000000000000000000000000000000003099077d1b17613b7dcd"}% 
```

**Reply by lingmicha** [reply]
This time it works!

Thanks a lot @davidmurdoch @tenthirtyone ！
